### ISSUE TITLE
[feature] Allow alerts to fire but not notify on first firing

### ISSUE THREAD

**Issue by cerebrate** [redstone99/hass-alert2#5]
I've recently discovered and started migrating over to Alert2 given the deprecation of HA-native Alert, and while everything's going well so far, I've found that we've got one usage pattern here that we haven't found a clean way to reproduce with Alert2.

We have some alerts (such as, for example, door left open alerts) which we used the _skip_first_ option for, intending that we should only be notified if the door was open for longer than the period of the alert. In setting this up on Alert2, we used the _delay_on_secs_ option to parallel this.

Unfortunately, when _deliberately_ leaving the door open for extended periods, we got into the habit of opening the door, then opening HA and acknowledging the alert preemptively before it notifies. Since with the Alert2 configuration the alert doesn't fire until after _delay_on_secs_, we can't do this.

So it would be very useful for us if there was a parallel option to have the alert fire (so we can ack it), but not trigger the notifiers until the reminder interval elapses at least once, or some other equivalent delay.

**Comment by redstone99** [comment]
Hi @cerebrate , thanks for detailed suggestion!  It sounds like you're describing a scenario where you want to indicate to Alert2 that the door being left open is not a problem for a period of time (because you're intentionally leaving it open).  Seems like here might be two possible ways of handling that:

If you want explicit control over marking the beginning and end of the interval, you could disable the alert and then re-enable the alert. You could make a UI button/toggle to trigger an automation to do that, or we could make it easier to do such a thing.

Alternatively, you could preemptively snooze the alert for a fixed time interval (and resnooze as necessary).  You could similarly make a UI button+automation to do that.  Snoozing would mean you'd get a summary notification at the end of the interval.  We could make a version of snooze that implicitly acks at the end of the snooze to avoid that.

Do either of those abstractions map well onto what you want to do?

Oh also, to be fair, I don't think Alert has been deprecated.  I named this project Alert2 aspirationally in the hope that one day it replaces Alert in HA.
J

**Comment by redstone99** [comment]
Thinking about this some more, it seems like there are two questions:

1. When the door is going to be propped open, do you want to snooze alert notification or prevent the alert from firing all together?  I can see advantages to either.
2. How do you want to indicate when the alert should return to normal?  It could be a manual step, or based on a timer expiring, or based on some fancy auto-detect logic

J

**Comment by cerebrate** [comment]
Thanks for the quick responses.

Let me go back a bit to cover our usage scenarios, since I wasn't as clear as I wanted to be.

There are pretty much three door scenarios:

1. Open it, go in/out, close it again.
2. Open it to go in and out, then forget about it and leave it open.
3. Deliberately open it and leave it open (for air, dogs, whatever).

(2) is pretty much the normal alert scenario. I set up `skip_first` initially to handle (1) - because it gets awfully noisy to have the alert-on and alert-off messages sounding every time we go in and out - and (3), because when we decided to leave the door open, we'd often remember to ack the alert before it ever sounded.

---

So given that, to answer the first of the latter questions, I'd say we just want to snooze notifications. We liked having the alert show as active on the dashboard whenever the door is open because that gives us just one place to look for a summary of things we should be aware of.

(Also, for later use cases, we may want to unsnooze/unack that alert via automation so that, for example, if we have left the door open into the evening or when the temperature drops significantly, it will start reminding us again.)

On the second, for the initial snooze, we'd want that to be timer-triggered (probably matching `reminder_frequency_mins`, to match the behavior of the old `skip_first`). After that, we'd just unack it manually.

That said, I've been becoming more familiar with (as I experiment) the Alert2 logic for snoozing, and I'm thinking this is going to be hard to implement simply, since as I've come to understand it, snoozing acks the alert, and as such, (a) it can't be manually acked while snoozed without first reenabling it, and (b) when reenabled or the snooze expires, the alert stays acked and so notifications don't resume?

Also, if I read this correctly, snoozing disables even new activations of the alert if it goes off and on again during the snooze period, so using the snooze login in this way might have one activation stomp on another accidentally?

---

On one final note - which I don't think Alert has been officially deprecated, there's [this thread](https://community.home-assistant.io/t/wth-are-alerts-not-configurable-through-the-ui/804195) on the Home Assistant community which links to some remarks by developers that Alert has been frozen and they're looking towards using automation blueprints to reproduce its functionality in future.

(I think that would be a horrible, _horrible_ way to do alerts, so I was very pleased indeed to find Alert2 as an alternative!)

**Comment by redstone99** [comment]
Thanks for the extra context.  And I didn't know Alert-1 was frozen - thanks for pointing me at that thread.

At a high level, an Alert2 entity currently has two main states - "off" and "on" where "on" means a problem that requires attention and remedy.  It sounds like you're interested in a third state, that of "something to be aware of", and you want to be able to switch from "something to be aware of" to "on" e.g. if the temperature drops and the door's still open.  And I think when you switch to "on" in that case, you want it to be as if the alert had been off and now just started firing.  Is that right?

Some possibilities:
1. My inclination is to add a new method, say `reset_alert`, that undoes snooze, ack, and the reminder timing logic. So after calling `reset_alert`, if the underlying condition is true, the alert will fire and notify as if it just started firing.

1. You could instead, unack and then unsnooze the alert, and you will start getting reminders again, though the timing will be based on where you are in the reminder timing sequence.

1. Another option is to more formally add a "something to be aware of" state, like maybe it's a boolean attribute.  You might be able to create a better UI experience if the system was aware of this as an explicit state.  I'm not sure of the specifics.

What do you think of option #1?  So you'd use snooze when you open the door, and then if the temp drops, you'd call reset_alert.
-Josh

**Comment by mitman98** [comment]
Joining the convo here - FWIW a discussion I was going to raise myself was alert escalation, i.e. first notify one notifier(s) then after a certain amount of time if the alert has not resolved, a second (e.g. escalated) notifier(s). If you just didn't put any notifiers on the first stage, you pretty much have #3...

Given that I think it's worth thinking out the generic feature set to hit a bit, but there might be a few birds that can be killed with one feature stone. So I'd personally avoid too narrow of an implementation as I think there's some really useful broader cases (both OP and mine and potentially others) that can be hit with the right (albeit similar) functionality.

**Comment by redstone99** [comment]
Let's talk about the broader escalation and related topics on the main HA forum thread: https://community.home-assistant.io/t/alert2-a-new-alerting-component/654597/98.   Interesting point about #3 and notifiers/escalation.

Re @cerebrate 's use case, my inclination is a reset_alert call, but let's see if the discussion about escalation suggests a different approach.

**Comment by redstone99** [comment]
@cerebrate , thinking a bit more about your use case.  So the door's open and you've snoozed the alert since you've propped the door open. Now the temperature drops.  In addition to unsnoozing/resettting the alert, I imagine you also would want the notification message to mention that a) the alert is restarting and b) the reason is that the temp has dropped.  I'm not sure how to think about that.

Having two distinct alerts, `door_open` and `door_open_temp_low` doesn't seem quite right.  But at the same time, when the temp drops, the alert circumstances have qualitatively changed.  And I can imagine other circumstance changes that might matter to the door_open alert:
* temp drops
* you leave the house, or go to sleep
* A/C turns on (temp rises)
* sun sets

Are these some kind of "sub-alert"?   One possibility is adding a `reason` string parameter to the hypothetical new call `reset_alert`.  Hrm.  You kinda want the reason to be visible in the UI as well, so maybe it needs to be part of the alert state.

Clearly I haven't figured this out yet.  I welcome any thoughts (cc @mitman98).

**Comment by redstone99** [comment]
I just posted on the main Alert2 thread some updated thoughts.

**Comment by cerebrate** [comment]
Sorry for the delay in getting back to you; busy time over the holidays.

Hrm.

At this point, while I'm really liking the proposals (I would have definite use for the proposed _reset_alert_ action, especially if it were allowed to reset a list of alerts and a generic "all alerts" as well as a single alert. I'm picturing, for example, after a specific time of day resetting all the door and window alerts so we can be sure to review them all before going to bed.) With regard to the sub-alerts, I like the "reason" string option, which I suppose would need to be added to the state of the alert until next off/ack/reset?

Where "something to be aware of" is concerned, as a state, that's pretty much how I think of on-but-acked anyway, which is something that's easy enough to pull out with a template using the _last_ack_time_ attribute.

To get back to the initial use case, though, unless I'm missing something, this would mean adding two automations to each alert that I wanted delayed notification on, which seems a little clunky, and the logic gets a bit complex when I want to manually (long-term) ack the alert before the delay period is up.

Since I do want the alert to fire immediately and be immediately ackable, ideally, just not to be notified immediately, I'm still thinking that tweaking something in the _reminder_frequency_mins_ logic might be the easiest way to get there, letting us specify a first item that is the delay before the notification fires? Changing the way it works is out for backwards compatibility, but there might be a way to flag the syntax, say starting with a negative number to indicate that it is the delay, so for my every-ten-minutes-but-wait-five-first configuration I used for the door, the config would look like:

```
- domain: front_door
  name: left_open
  friendly_name: "Front Door Left Open"
  condition: binary_sensor.front_door
  reminder_frequency_mins: [-5, 10]
  message: "The front door has been left open."
  done_message: "The front door has been closed."
```

**Comment by mitman98** [comment]
@cerebrate please look at the HA Community Alert2 thread. We're discussing staged alerts that allows for escalated notifications that is being designed in a way that should be solving your use case, but also generically addresses other various paths of escalated states/notifications. It has more updated info that the comments on this thread.

**Comment by cerebrate** [comment]
@mitman98 Yep, that was my next port of call.

So, I looked at your example, and tried to solve my use case with it, and came up with this:

```
- domain: front_door
  name: left_open
  friendly_name: "Front Door Left Open"
  condition: binary_sensor.front_door
  notifier: []
  escalation_stages:
    1:
      condition: binary_sensor.front_door
      delay_on_secs: 300
      message: "The front door has been left open."
      done_message: "The front door has been closed."
      notifier: standard_persistent
```

Which is a bit wordy, but hey, if it works I won't complain. :)

But a couple of things did occur to me:

- If I ack the alert before the 5 minutes-to-escalation is up, wouldn't the escalation unack the alert and play the notification? In the example you give with a threshold alert, and in the general case of escalations, this would be desirable behavior, so I would think it would work that way.
- As a minor point, I'd like to get the `done_message` even if I ack it before it escalates, so it would be good to have that in the base alert config, but that means I couldn't use `notifier: []` to silence the pre-escalation alert, which means we'd need a way to specify an empty message and have that mean "don't notify at all".

That said, I love the staged alert concept and would definitely use it, but I don't think it's quite sufficient/perfectly applicable here.

**Comment by mitman98** [comment]
@cerebrate 

For the YAML itself I don't think we've solidified it's exact semantics and defaults. I can imagine that you don't have to specify condition a second time. But honestly you're getting into some corner cases I don't think there's going to be a perfect solution that's compact without Alert2 special casing itself to this use case which doesn't seem architecturally sound if we're getting constructs that get there without needing automations, helpers, etc.

For your first bullet i suspect the desired behavior is that ack would be unacked on escalation, that makes sense to me. But isn't that desirable in your use case? 

I do think the corner cases of message, done_message, notifications, etc. are all completely thought out yet with stages, this is still fairly early in concept. 

Could you elaborate on where you feel the design isn't sufficient for your use case? That'll help with the design.

It sounds like you have a particular use case where you want an alert to show up in the UI but not be notified when it happens; get an alert if a certain amount of time has passed; but also get a done_message even if the time hasn't passed and the original alert ended? I.e. the desire to have no notification when the alert happens but a done_message when it's done? That sounds getting a little bit special cased.

A thought just came to me, maybe instead of trying to force alert2 to do all this in its code we allow Alert2 to also call services (with good arguments containing useful information) and not just fire notifications when changes occur. THen you can write a script to do whatever you wanted for your particular case that's unique... doesn't force Alert2 to have to solve every problem itself. That would be a powerful feature.

**Comment by cerebrate** [comment]
Let me go right back to my use case, 'cause I think we may have obscured the details over the course of discussion. Basically, I'm trying to emulate the behavior of the old Alert integration with _skip_first_ set in two particular ways. (Only one of which was mentioned in this issue originally, to be clear.) I'm going to use door open alerts as my example, because it's my most common usage of these things.

1. _skip_first_ meant that I would not be notified on "short" door openings and closings. This is useful because I don't want to get open and close announcements every time I just pass through the door. But unlike using _delay_on_secs_ in Alert2, the alert still _fires_ immediately, it just doesn't notify immediately. This means that the alerts panel in the UI gives consistent answers on whether the door is open or not¹, and also means that if I decide to leave the door open, I can ack it immediately without having to wait for it to fire/notify me the first time².
 
(This is why ack being unacked on escalation doesn't fit this use case. If I ack it preemptively, I want it to stay acked _in this case_. Which is awkward for staged alerts as an implementation method, since unacking is correct for just about every other use case.)

Which may sound complicated when I try to spell it out, but really, I'm just asking for _skip_first_ or, rather, the _delay_notify_secs_ counterpart to _delay_on_secs_ so that I can see transients, but not be bugged about transients.

2. (the one I didn't talk about here because it's not really part of the same issue) I would like have the option to receive the `done_message` for acked (condition) alerts that go to off, for reasons that are mostly about multiple users. If someone else closes the door or does whatever the needful is to clear the alert, it's helpful to be told that.

Where these interact with staged alerts - well, the way I see it, if I can implement what I want using staged alerts, then I'll do that, but in my own opinion, it's a poor fit for it. There are lots of uses I'd have for them (and I have some thoughts on that I'll post over on the community thread once they fully gel), but they'd be a baroque way to implement something as conceptually simple as _delay_notify_secs_, no?

---

1. I can also see the state of the door sensors directly, of course, but the alerts panel is intentionally right there when I open up HA. It's hard to change longstanding habits, especially since they're reinforced every time a door _is_ left open.

2. If I decide to leave it open, I don't want to be nagged about it. (I have automations that will unack certain notifications if situations change, which would be a great use for staged alerts when/if they happen.) But I do want to have the alert fire, so when I open HA I get a complete list of things that includes those I've chosen to ignore for now. Makes sense?

**Comment by mitman98** [comment]
Thank you for that detailed and useful description that was very helpful! I’ll continue to think now with that info.

**Comment by redstone99** [comment]
@cerebrate, great summary.  I consider this project in it's infancy, so prioritize getting to a design that we like and not being too constrained by how things are currently.  There are a few ways we can manage changes that might conflict with backward compatibility.  To your use case, yeah, I'd like a convenient solution that doesn't require you to add extra automations on top of Alert2 (within reason).

I think there are a few possible solutions.  It'll be easier for me to dig in once I'm done with the current UI work.
-J

**Comment by redstone99** [comment]
I just posted a proposal for staged alerts on the main thread:
https://community.home-assistant.io/t/alert2-a-new-alerting-component/654597/144

**Comment by redstone99** [comment]
I think we can consider this feature done and I'm inclined to close this unless anyone objects.
